### PR TITLE
Tree isomorphism

### PR DESCRIPTION
**Pull request by kosii**
For #1602 . I'd just like to have an early feedback, I know it still misses handling eventual corner cases, tests and docs.

### CODE REVIEW

File: networkx/algorithms/isomorphism/tree_isomorph.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        raise NetworkXError("Graph {} is not a tree.".format(T1))
+
+    if not nx.is_tree(T2):
+        raise NetworkXError("Graph {} is not a tree.".format(T2))
+
+    return canonical_form(T1, r1) == canonical_form(T2, r2)
+
+
+def tree_is_isomorphic(T1, T2):
+    if not T2:
```
Comment (jfinkels) [inline networkx/algorithms/isomorphism/tree_isomorph.py]: Might be more readable as `if not T1 and not T2: return True`.

File: networkx/algorithms/isomorphism/tree_isomorph.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        return not T1
+
+    r2 = nx.center(T2)[0]
+    for r1 in nx.center(T1):
+        if rooted_tree_is_isomorphic(T1, T2, r1, r2): 
+            return True
+    return False
```
Comment (jfinkels) [inline networkx/algorithms/isomorphism/tree_isomorph.py]: Why not just `return any(rooted_is_isomorphic(T1, T2, r1, r2) for r1 in centers1 for r2 in centers2)`?

File: networkx/algorithms/isomorphism/tree_isomorph.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    if not nx.is_tree(T1):
+        raise NetworkXError("Graph {} is not a tree.".format(T1))
+
+    if not nx.is_tree(T2):
+        raise NetworkXError("Graph {} is not a tree.".format(T2))
+
+    return canonical_form(T1, r1) == canonical_form(T2, r2)
+
+
+def tree_is_isomorphic(T1, T2):
```
Comment (jfinkels) [inline networkx/algorithms/isomorphism/tree_isomorph.py]: I think it would be better to have these as a single function, `is_isomorphic()` in the module `isomorphism/tree.py`, with keyword arguments for the (optional) roots:

    def is_isomorphic(T1, T2, root1=None, root2=None):
        if root1 is not None and root2 is not None:
            return canonical_form(...) == ...
        if root1 is None and root2 is None:
            return any(is_isomorphic(T1, T2, r1, r2) for r1 in center(T1) for r2 in center(T2)))
        raise Exception

File: networkx/algorithms/isomorphism/tree_isomorph.py (hunk 2 above)
Comment (dschult) [inline networkx/algorithms/isomorphism/tree_isomorph.py]: This suggestion looks harder to read...  
Python indentation is much of why python is so readable.... let's use it.

File: networkx/algorithms/isomorphism/tree_isomorph.py (hunk 1 above)
Comment (dschult) [inline networkx/algorithms/isomorphism/tree_isomorph.py]: But the suggestion is not the same meaning....
Original PR returns False but suggestion only returns True.

File: networkx/algorithms/isomorphism/tree_isomorph.py (hunk 2 above)
Comment (jfinkels) [inline networkx/algorithms/isomorphism/tree_isomorph.py]: Yeah, I am suggesting my personal preference. But I prefer to use the functional `any` (and `all`) where possible because 1. it abstracts the common procedural pattern

    for x in foo:
        if condition(x):
            return True
    return False

so there is less code to write and therefore fewer opportunities for logic errors, 2. the name matches our English language description of that pattern, and 3. it makes it clear where parallelism can be applied: wherever you see an `any` or `all`!

File: networkx/algorithms/isomorphism/tree_isomorph.py (hunk 1 above)
Comment (jfinkels) [inline networkx/algorithms/isomorphism/tree_isomorph.py]: I thought the intention (not the implementation) was that when both are empty, immediately return `True`, otherwise run the algorithm.

Comment (kosii) [inline networkx/algorithms/isomorphism/tree_isomorph.py]: It was to handle the case when T2 is empty (and so there is no center).
Should two empty graphs be considered as empty? Because nx.is_tree(nx.Graph()) raises NetworkXError.

File: networkx/algorithms/isomorphism/tree_isomorph.py (hunk 3 above)
Comment (kosii) [inline networkx/algorithms/isomorphism/tree_isomorph.py]: Good idea

File: networkx/algorithms/isomorphism/tree_isomorph.py (hunk 1 above)
Comment (jfinkels) [inline networkx/algorithms/isomorphism/tree_isomorph.py]: Hm, I believe that two null graphs should be considered isomorphic. The definition of isomorphic for graphs with node sets *U* and *V* is "there exists a function *f* from *U* to *V* such that *f* is a bijection and *f* preserves edges". Since both *U* and *V* are empty, there is just one function from *U* to *V*, the [empty function][1], and [it is a bijection][2]. Furthermore, the edge preservation requirement is [vacuously true][3].

However, the behavior of the tree isomorphism function should be "undefined" for non-tree inputs (that is, anything can happen). I think that, as with the `bipartite` package, the responsibility is on the calling code to decide whether the given graphs are actually trees before calling this `is_isomorphic` function. In other words, a pre-condition for calling this function will be that each graph is a tree. Since an empty graph is not a tree, you can assume that you will always get non-empty graphs as input.

[1]: https://en.wikipedia.org/wiki/Empty_function
[2]: http://math.stackexchange.com/questions/562412/is-the-empty-function-always-a-bijection
[3]: https://en.wikipedia.org/wiki/Vacuous_truth
